src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,414 @@
+            labelled (bool):
+                If True, returns the xarray with labelled dimensions,
+                overwise returns numpy.ndarray
+
+        Returns:
+            positions (numpy.ndarray or xarray.DataArray):
+                The motor positions
+        """
+        train_ids, pos, counts = self._read_positions()
+        pos = np.repeat(pos, counts, axis=0)
+
+        if labelled:
+            dims = ["trainId"] + list(self.dims.keys())
+            coords = {"trainId": self.train_ids}
+            coords.update({name: values for name, values in self.dims.items()})
+            return xarray.DataArray(
+                pos, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return pos
+
+    def _read_positions(self):
+        """Reads and compresses motor positions."""
+        if hasattr(self, "_positions"):
+            return self._positions
+
+        # read positions
+        pos = np.zeros((self.num_trains, self.num_sources), dtype=float)
+        for source_no, key_data in enumerate(self.keys):
+            pos[:, source_no] = key_data.ndarray()
+        pos = pos.reshape(-1, *self.shape)
+
+        # compress
+        axes = tuple(range(1, pos.ndim))
+        ix = np.flatnonzero(
+            np.insert(np.any(np.diff(pos, axis=0), axis=axes), 0, True))
+
+        # train ids when positions are changed
+        train_ids = np.asarray(self.dc.train_ids)[ix]
+        # count of trains at every position
+        counts = np.diff(np.append(train_ids.astype(int),
+                                   int(self.dc.train_ids[-1]) + 1))
+
+        self._positions = train_ids, pos[ix], counts
+        return self._positions
+
+    def compressed_positions(self, labelled=True):
+        """Returns the unique motor positions and corresponding train IDs.
+
+        Args:
+            labelled (bool):
+                If True, returns the xarray with labelled dimensions,
+                overwise returns tuple of numpy.ndarrays
+
+        Returns:
+            positions (tuple of two numpy.ndarray or xarray.DataArray):
+                The motor positions
+        """
+        train_ids, values, _ = self._read_positions()
+        if labelled:
+            dims = ["trainId"] + list(self.dims.keys())
+            coords = {"trainId": train_ids}
+            coords.update({name: values for name, values in self.dims.items()})
+            return xarray.DataArray(
+                values, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return train_ids, values
+
+    def positions_at(self, tid):
+        """Returns motor positions at given train.
+
+        Args:
+            tid (int):
+                Train ID
+
+        Returns:
+            postions (numpy.ndarray):
+                The motor positions
+        """
+        train_ids, values, _ = self._read_positions()
+        i = np.searchsorted(train_ids, tid, side="right")
+        return values[max(i - 1, 0)]
```

Comment:
I think this should throw an error if the `tid` is not in `self.train_ids`.

Reply:
Good catch!